js/eventloop.go

```diff
@@ -0,0 +1,165 @@
+// This function *must* be called from within running on the event loop, but it's result can be called from anywhere
+// TODO better name
+func (e *eventLoop) reserve() func(func() error) {
+	e.lock.Lock()
+	e.reservedCount++
+	e.lock.Unlock()
+
+	return func(f func() error) {
+		e.lock.Lock()
+		e.queue = append(e.queue, f)
+		e.reservedCount--
+		e.lock.Unlock()
+		e.wakeup()
+	}
+}
+
+func (e *eventLoop) promiseRejectionTracker(p *goja.Promise, op goja.PromiseRejectionOperation) {
+	// No locking here as the goja runtime will call this synchrnously
+	// Read Notes on https://tc39.es/ecma262/#sec-host-promise-rejection-tracker
+	if op == goja.PromiseRejectionReject {
+		e.pendingPromiseRejections[p] = struct{}{}
+	} else { // goja.PromiseRejectionHandle so a promise that was previously rejected without handler now got one
+		delete(e.pendingPromiseRejections, p)
+	}
+}
+
+func (e *eventLoop) popAll() (queue []func() error, reserved bool) {
+	e.lock.Lock()
+	queue = e.queue
+	e.queue = make([]func() error, 0, len(queue))
+	reserved = e.reservedCount != 0
+	e.lock.Unlock()
+	return
+}
+
+// start will run the event loop until it's empty and there are no reserved spots
+// or a queued function returns an error. The provided function will be the first thing executed.
+// After start returns the event loop should not be reused if an error was returend
+func (e *eventLoop) start(f func() error) error {
+	e.queue = []func() error{f}
+	for {
+		queue, reserved := e.popAll()
+
+		if len(queue) == 0 {
+			if !reserved { // we have empty queue and nothing that reserved a spot
+				return nil
+			}
+			<-e.wakeupCh // wait until the reserved is done
+			continue
+		}
+
+		for _, f := range queue {
+			if err := f(); err != nil {
+				return err
+			}
+		}
+
+		// yes it will be at random but that seems to be the case in other tools as well
+		for promise := range e.pendingPromiseRejections {
+			// TODO maybe throw the whole promise up and get make a better message outside of the event loop
+			value := promise.Result()
+			// try to get the stack without actually needing the runtime
+			// this might break in the future :(
+			if o, ok := promise.Result().(*goja.Object); ok {
+				stack := o.Get("stack")
+				if stack != nil {
+					value = stack
+				}
+			}
+			// this is the de facto wording in both firefox and deno at least
+			return fmt.Errorf("Uncaught (in promise) %s", value) //nolint:stylecheck
+		}
+	}
+}
+
+// this exists so we can wait on all reserved so we know nothing is still doing work
+func (e *eventLoop) waitOnReserved() {
+	for {
+		_, reserved := e.popAll()
+		if !reserved {
+			return
+		}
+		<-e.wakeupCh // wait until the reserved is done
+	}
+}
+
+func (e *eventLoop) addSetTimeout(ctx context.Context, rt *goja.Runtime) {
+	_ = rt.Set("setTimeout", func(f func() error, t float64) {
+		// TODO checks and fixes
+		// TODO maybe really return something to use with `clearTimeout
```

Comment:
I prefer to do in a subsequent PR, possibly still in this release, but not actually add more stuff to this PR. On top of that `setTimeout` might be removed before merging if enough people thing that we shouldn't have it in the first place.

Reply:
The `setTimeout` should [return an ID](https://developer.mozilla.org/en-US/docs/Web/API/setTimeout#return_value) for passing it to the `clearTimeout`. So +1 for me to have `setTimeout` in a dedicated PR, possibly in the same release.

Reply:
:+1: `setTimeout` + `clearTimeout` should be implemented together, both to follow the APIs but also to know there aren't any implementation obstacles. However, I am fine with both a separate v0.37.0 PR on top of this one with them, or a separate commit in this PR...

The problem with moving them to a separate PR is that this PR will be left pretty "naked". You'll have to remove the HTTP changes for sure, so if you also remove `setTimeout`, what will be left besides the raw even loop, nothing to test it with? :sweat_smile:

Reply:
see https://github.com/grafana/k6/pull/2228#issuecomment-1047747652

I am somewhat against shipping these by default, but if we do, maybe there some TC39 tests that we can enable? :thinking: :man_shrugging:

Reply:
There are some tc39 tests but I think it will be a lot of work to make them work with the event loop, big enough that I don't think it makes sense for the MVP. And likely most of those will require `clearTimeout` .